### PR TITLE
Require trait properties in MiniYaml

### PR DESCRIPTION
**Pull request by atlimit8**
Allows for marking certain trait properties as required using `[FieldLoader.Require]` or `[FieldLoader.LoadUsing("...", true)]`.

Based on comment from @Phrohdoh on #8698.

To test, just comment out one of the required trait properties.

### CODE REVIEW

**Comment by pchote** [comment]
This is a great idea.  The implementation looks sensible, but I haven't gone over the changes line by line and can't run the game to test anything.

**Comment by pchote** [comment]
The [trait docs](https://github.com/OpenRA/OpenRA/wiki/Traits) will need to be updated to reflect this.  I suggest replacing the default value (which is meaningless) with something like **[Required]** for these properties.

**Comment by atlimit8** [comment]
That should be done by the changes to `OpenRA.Mods.Common/UtilityCommands/ExtractTraitDocsCommand.cs`, except its more like *(required)*.

**Comment by Phrohdoh** [comment]
What if something is required but the default value is the desired value?

**Comment by atlimit8** [comment]
That is a contradiction. Either the MiniYaml must overwrite the default making it meaningless, or the MiniYaml is optional and the default value is used if the MiniYaml is not present.

**Comment by atlimit8** [comment]
`[FieldLoader.Require]` and `[FieldLoader.LoadUsing("...", true)]` force the default value to be overwritten or throw an exception stopping the engine. Either way, the hard coded value is not used.

I may be misunderstanding your question or you may misunderstand what `require` means in this context,

**Comment by pchote** [comment]
Oh sorry, when I skimmed through the code I completely missed that you had already done it.

**Comment by obrakmann** [comment]
Looks good to me and works as expected, :+1: 

I guess we will fine-tune these required annotations as we come across them. I think there are more candidates for it than this PR already adds (`Queue` in `Buildable` comes to mind, or `PanelName` in `ObjectivesPanel`.)

File: OpenRA.Game/GameRules/ActorInfo.cs [hunk 1]
```diff
@@ -94,7 +103,16 @@ static ITraitInfo LoadTraitInfo(string traitName, MiniYaml my)
...
 			var info = Game.CreateObject<ITraitInfo>(traitName + "Info");
-			FieldLoader.Load(info, my);
+			try
+			{
+				FieldLoader.Load(info, my);
+			}
+			catch (FieldLoader.MissingFieldsException e)
+			{
+				throw new FieldLoader.MissingFieldsException("Trait name " + traitName + ": "
+					+ (e.Message.Contains(',') ? "Required properties missing: " : "Required properties missing: ") + e.Message);
```
Comment (pchote) [inline OpenRA.Game/GameRules/ActorInfo.cs]: Typo: that second branch should be "Required property missing: ".

File: OpenRA.Game/FieldLoader.cs [hunk 2]
```diff
@@ -52,19 +58,33 @@ public static void Load(object self, MiniYaml my)
...
+							missing.Add(fli);
 						continue;
+					}
 				}
 
 				fli.Field.SetValue(self, val);
 			}
+
+			if (missing.Any())
+				throw new MissingFieldsException(missing.First().YamlName + string.Concat(missing.Skip(1).Select(r => ", " + r.YamlName)));
```
Comment (pchote) [inline OpenRA.Game/FieldLoader.cs]: If you added a `string[]` property for the missing fields then you won't need to detect on `,` when handling it (which is a bit gross).

**Comment by atlimit8** [comment]
I know I tried some more that looked like they should be required, but `make test` would fail. I decided it would be better for them to be handled by those more familiar and/or while those traits are being changed.

File: OpenRA.Game/FieldLoader.cs (hunk 2 above)
Comment (atlimit8) [inline OpenRA.Game/FieldLoader.cs]: Yea, I knew it was a hack. OK.

**Comment by atlimit8** [comment]
`FieldLoader.MissingFieldsException` now takes a `string[]` of missing properties and an optional header `string`.

**Comment by abcdefg30** [comment]
Looks good to me. :+1:

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d3281d33d1c34de58e4e4381c0700e2e063f6484...4b081ee9775085e9f34c9d5d55afe038ccf6f44f)

**Pull request merged**